### ISSUE TITLE
Regex mismatch between expected result from regex101 and Logos

### ISSUE THREAD

**Issue by sondr3** [maciejhirsz/logos#54]
I was asking around in the beginner channel on Discord for Rust with some help for a minor problem I had where the result from both regexr and regex101 doesn't match what a test I have does. Specifically, this regex: `[+-]?[0-9]*[.]?[0-9]+(?:[eE][+-]?[0-9]+)?` on both [RegExr](https://regexr.com/44mvc) and [regex101](https://regex101.com/r/If29xU/1/) show that it matches numbers with an `e` in them, but not `em`. However, in my Code I have 

```rust
    #[regex = "em|ex|ch|rem|vw|vh|vmin|vmax"]
    RelativeLength,
    #[regex = "[+-]?[0-9]*[.]?[0-9]+(?:[eE][+-]?[0-9]+)?"]
    Number,
```

And a test

```rust
    #[test]
    fn relative_lengths() {
        assert_lex(".container { width: 40em; } /* em  */ ",
        &[
            (Token::Period, "."),
            (Token::Ident, "container"),
            (Token::CurlyBracketOpen, "{"),
            (Token::Ident, "width"),
            (Token::Colon, ":"),
            (Token::Number, "40"),
            (Token::RelativeLength, "em"),
            (Token::CurlyBracketClose, "}"),
        ]);
    }
```

that outputs

```text
failures:

---- css::test::relative_lengths stdout ----
thread 'css::test::relative_lengths' panicked at 'assertion failed: `(left == right)`
  left: `(UnexpectedToken, "40e")`,
 right: `(Number, "40")`', src/css.rs:122:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

where you can see that it matches the `e` at the end of `40`. We were going back and forth on whether there was an issue with my regex, or `regex-syntax` and whatnot, but in the I was advised to ask here first. If you want a first hand look at this, my repo for this lives here: https://github.com/sondr3/minifie-rs.

Thanks for a great crate though! I'm having a lot of fun working with it 😃

**Comment by maciejhirsz** [comment]
Hey! Thanks, it's interesting for me to see how people use the crate in ways I never expected :).

So the issue here is that Logos, unlike regular Regex engines, never backtracks. In your case, it will read `40e` as a beginning of a number variant, then expects `+|-|0-9` on the next byte, and when it doesn't find it, it assumes a badly formatted number. #52 helps with some cases of this but not all. I guess I'll have to bite the bullet and add backtracking in for corner cases like this.

**Comment by sondr3** [comment]
Ah, yeah, we had a few back-and-forth on whether or not it was a problem with greediness or backtracking or a bug in the regex crate. Thanks for all the help!

**Comment by maciejhirsz** [comment]
So I figured a way to do this without backtracking _as such_. This is already passing my tests in #57 but there is still a bug I need to track and some low hanging fruit I want to add. Given two definitions: `"0x[0-9a-fA-F]+"` and `"[0-9]+"`, when encountering a string such as `"  0x  "` once `0` is read, the lexer figures out that the branch following `'x'` requires at least one another byte matching the following pattern to be valid, so it performs a bound check for the next 2 bytes, then checks that they are matching `"x[0-9a-fA-F]"` and only then bumps the lexer by `2`. If the `x` is not followed by anything it figures that `0` was a valid token and produces that instead.

TL;DR: [this test](https://github.com/maciejhirsz/logos/blob/a6bbc860e376e2f336862e3d03120437ca5ddd81/tests/tests/simple.rs#L261-L269) is now passing:

```rust
    #[test]
    fn invalid_hex_and_binary() {
        assert_lex("0x 0b", &[
            (Token::Number, "0", 0..1),
            (Token::Identifier, "x", 1..2),
            (Token::Number, "0", 3..4),
            (Token::Identifier, "b", 4..5),
        ]);
    }
```

There are limits to how far it can lookahead, so it's not proper backtracking that can undo entire complex branches, but it should cover most cases where backtracking would necessary without running explosive performance drops that backtracking tends to introduce.
